### PR TITLE
Refactor Pyomo Dispatch

### PR DESCRIPTION
**Pull request by dylanjm**
--------
Pull Request Description
--------
##### What issue does this change request address?

#332

##### What are the significant changes in functionality due to this change request?

+ New Files: `PyomoRuleLibrary.py` `PyomoModelHandler.py` `putils.py`

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/HERON/wiki/Code-Standards) for details.
- [x] 4. Automated Tests should pass.
- [x] 5. If significant functionality is added, there must be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large tes.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added, the the analytic documentation must be updated/added.
- [x] 9. If any test used as a basis for documentation examples have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: src/dispatch/PyomoModelHandler.py [hunk 1]
```diff
@@ -0,0 +1,576 @@
...
+      self._process_storage_component(component, interaction)
+    else:
+      activity = interaction.get_strategy().evaluate(self.meta)[0]['level']
+      self._create_production_param(component, activity)
+
+
+  def _process_storage_component(self, component, interaction):
+    """
+      Process a storage component.
+      @ In, m, pyo.ConcreteModel, associated model
```
Comment (PaulTalbot-INL) [inline src/dispatch/PyomoModelHandler.py]: remove `m` from the arg list?

File: src/dispatch/PyomoModelHandler.py [hunk 2]
```diff
@@ -0,0 +1,576 @@
...
+    charge = np.where(deltas > 0, -deltas / dt / rte2, 0)
+    discharge = np.where(deltas < 0, -deltas / dt * rte2, 0)
+    self._create_production_param(component, charge, tag="charge")
+    self._create_production_param(component, discharge, tag="discharge")
+
+
+  def _create_production_limit(self, validation):
+    """
+      Creates pyomo production constraint given validation errors
+      @ In, m, pyo.ConcreteModel, associated model
```
Comment (PaulTalbot-INL) [inline src/dispatch/PyomoModelHandler.py]: remove `m`

Comment (PaulTalbot-INL) [reply src/dispatch/PyomoModelHandler.py]: also for many methods in this file, I'll stop tagging them individually.

File: src/dispatch/PyomoModelHandler.py [hunk 3]
```diff
@@ -0,0 +1,576 @@
...
+    setattr(self.model, prod_name, prod)
+    return prod_name
+
+
+  def _create_production(self, comp):
+    """
+      Creates all pyomo variable objects for a non-storage component
+      @ In, m, pyo.ConcreteModel, associated model
+      @ In, comp, HERON Component, component to make production variables for
+      @ In, meta, dict, dictionary of state variables
```
Comment (PaulTalbot-INL) [inline src/dispatch/PyomoModelHandler.py]: remove `meta`

File: src/dispatch/PyomoModelHandler.py [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+      setattr(self.model, f'{comp.name}_ramp_freq_constr', constr)
+
+
+  def _create_capacity_constraints(self, comp, prod_name):
+    """
+      Creates pyomo capacity constraints
+      @ In, m, pyo.ConcreteModel, associated model
+      @ In, comp, HERON Component, component to make variables for
+      @ In, prod_name, str, name of production variable
+      @ In, meta, dict, additional state information
```
Comment (PaulTalbot-INL) [inline src/dispatch/PyomoModelHandler.py]: `m`, `meta`

File: src/dispatch/PyomoModelHandler.py [hunk 5]
```diff
@@ -0,0 +1,576 @@
...
+
+
+  def _create_storage(self, comp):
+    """
+      Creates storage pyomo variable objects for a storage component
+      Similar to create_production, but for storages
+      @ In, m, pyo.ConcreteModel, associated model
+      @ In, comp, HERON Component, component to make production variables for
+      @ In, initial_storage, dict, initial storage levels
+      @ In, meta, dict, additional state information
```
Comment (PaulTalbot-INL) [inline src/dispatch/PyomoModelHandler.py]: fix arg list

File: src/dispatch/PyomoRuleLibrary.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+      return delta <= limit * up - eps * down
+
+def ramp_freq_rule(Bd, Bu, tao, t, m) -> bool:
+  """
+    Constructs pyomo frequency-of-ramp constraints.
+    @ In, Bd, bool var, binary tracking down-ramp events
+    @ In, Bu, bool var, binary tracking up-ramp events
+    @ In, tao, int, number of time steps to look back
+    @ In, t, int, time step indexer
+    @ In, m, pyo.ConcreteModel, pyomo model
```
Comment (PaulTalbot-INL) [inline src/dispatch/PyomoRuleLibrary.py:204]: needs a `Out`

File: src/dispatch/PyomoRuleLibrary.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+def ramp_rule_up(prod_name, r, limit, neg_cap, t, m, bins=None) -> bool:
+  """
+    Constructs pyomo production ramping constraints.
+    @ In, prod_name, str, name of production variable
+    @ In, r, int, index of resource for capacity constraining
+    @ In, limit, float, limiting change in production level across time steps
+    @ In, neg_cap, bool, True if capacity is expressed as negative (consumer)
+    @ In, t, int, time index for ramp limit rule (NOTE not pyomo index, rather fixed index)
+    @ In, m, pyo.ConcreteModel, associated model
+    @ In, bins, tuple, optional, (lower, steady, upper) binaries if limiting ramp frequency
```
Comment (PaulTalbot-INL) [inline src/dispatch/PyomoRuleLibrary.py:167]: Needs `Out`

File: src/dispatch/PyomoRuleLibrary.py [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+
+def ramp_freq_bins_rule(Bd, Bu, Bn, t, m) -> bool:
+  """
+    Constructs pyomo constraint for ramping event tracking variables.
+    This forces choosing between ramping up, ramping down, and steady state operation.
+    @ In, Bd, bool var, binary tracking down-ramp events
+    @ In, Bu, bool var, binary tracking up-ramp events
+    @ In, Bn, bool var, binary tracking no-ramp events
+    @ In, t, int, time step indexer
+    @ In, m, pyo.ConcreteModel, pyomo model
```
Comment (PaulTalbot-INL) [inline src/dispatch/PyomoRuleLibrary.py:225]: `Out`

File: src/dispatch/PyomoRuleLibrary.py [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+  return total
+
+def conservation_rule(res, m, t) -> bool:
+  """
+    Constructs conservation constraints.
+    @ In, initial_storage, dict, initial storage levels at t==0 (not t+offset==0)
+    @ In, res, str, name of resource
+    @ In, m, pyo.ConcreteModel, associated model
+    @ In, t, int, index of time variable
+    @ Out, conservation, bool, balance check
```
Comment (PaulTalbot-INL) [inline src/dispatch/PyomoRuleLibrary.py:249]: fix arg list

File: src/dispatch/putils.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+  # # TODO minimum!
+  # # producing or consuming the defining resource?
+  # # -> put these in dictionaries so we can "get" limits or return None
+  # if maximum > 0:
+  #   lower = {r: 0}
+  #   upper = {r: maximum}
+  # else:
+  #   lower = {r: maximum}
+  #   upper = {r: 0}
+  # return lower, upper
```
Comment (PaulTalbot-INL) [inline src/dispatch/putils.py]: I think we can just remove this method

**Review by PaulTalbot-INL — COMMENTED** [review]
A few comments to address, I think they're all docstrings related tbh.

File: src/dispatch/PyomoRuleLibrary.py (hunk 7 above)
Comment (dylanjm) [reply src/dispatch/PyomoRuleLibrary.py:167]: Fixed

File: src/dispatch/PyomoModelHandler.py [hunk 11]
```diff
@@ -0,0 +1,558 @@
...
+    prod = pyo.Param(res_indexer, self.model.T, initialize=dict(init))
+    setattr(self.model, prod_name, prod)
+    return prod_name
+
+
+  def _create_production(self, comp):
+    """
+      Creates all pyomo variable objects for a non-storage component
+      @ In, comp, HERON Component, component to make production variables for
+      @ Out, None
```
Comment (PaulTalbot-INL) [inline src/dispatch/PyomoModelHandler.py]: this return out needs to be updated

File: src/dispatch/PyomoRuleLibrary.py [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+
+def prod_limit_rule(prod_name, r, limits, kind, t, m) -> bool:
+  """
+    Constructs pyomo production constraints.
+    @ In, prod_name, str, name of production variable
+    @ In, r, int, index of resource for capacity constraining
+    @ In, limits, list(float), values in time at which to constrain resource production
+    @ In, kind, str, either 'upper' or 'lower' for limiting production
+    @ In, t, int, time index for production rule (NOTE not pyomo index, rather fixed index)
+    @ In, m, pyo.ConcreteModel, associated model
```
Comment (PaulTalbot-INL) [inline src/dispatch/PyomoRuleLibrary.py:102]: missing out

**Review by PaulTalbot-INL — COMMENTED** [review]
Two last docstring fixes

**Review by PaulTalbot-INL — APPROVED** [review]

**Comment by PaulTalbot-INL** [comment]
Code reviewed and tests passed on most recent code changes (last update is just docstring fixes). Merging.

**Pull request merged**